docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1059 @@
+"valid sequence" has passed a particular list, then that list may
+be executed.
+
+Deletion queue files will be stored in a sibling of the `tenants/` directory, and
+with the node generation number in the name. The paths would look something like:
+
+```bash
+  # Deletion List
+  deletion/<node id>/<sequence>-<generation>.list
+
+  # Header object, stores the highest sequence & validated sequence
+  deletion/<node id>/header-<generation>
+```
+
+Each entry in a deletion list is structured to contain the tenant & timeline,
+and the attachment generation.
+
+```rust
+/// One of these per deletion list object
+struct DeletionList {
+  deletions: Vec<DeletionEntry>
+}
+
+/// N of these per deletion list
+struct DeletionEntry {
+  tenant_id: TenantId,
+  timeline_id: TimelineId,
+  generation: AttachmentGeneration
+  keys: Vec<String>,
+}
+```
+
+#### Operations that may skip the queue
+
+Deletions of an entire timeline are exempt from generation number validation. Once the
+control plane sends the deletion request, there is no requirement to retain the readability
+of any data within the timeline, and all objects within the timeline path may be deleted
+at any time from the control plane's deletion request onwards.
+
+Since deletions of smaller timelines won't have enough objects to compose a full sized
+DeleteObjects request, it is still useful to send these through the last part of the
+deletion pipeline to coalesce with other executing deletions: to enable this, the
+deletion queue should expose two input channels: one for deletions that must be
+processed in a generation-aware way, and a fast path for timeline deletions, where
+that fast path may skip validation and the persistent queue.
+
+### Cleaning up orphan objects (scrubbing)
+
+An orphan object is any object which is no longer referenced by a running node or by metadata.
+
+Examples of how orphan objects arise:
+
+- A node is doing compaction and writes a layer object, then crashes before it writes the
+  index_part.json that references that layer.
+- A partition node carries on running for some time, and writes out an unbounded number of
+  objects while it believes itself to be the rightful writer for a tenant.
+- A pageserver crashes between un-linking an object from the index, and persisting
+  the object to its deletion queue.
+
+Orphan objects are functionally harmless, but have a small cost due to S3 capacity consumed. We
+may clean them up at some time in the future, but doing a ListObjectsv2 operation and cross
+referencing with the latest metadata to identify objects which are not referenced.
+
+Scrubbing will be done only by an attached pageserver (not some third party process), and deletions requested during scrub will go through the same
+validation as all other deletions: the attachment generation must be
+fresh. This avoids the possibility of a stale pageserver incorrectly
+thinking than an object written by a newer generation is stale, and deleting
+it.
```

Comment:
The deletion queue contains entries of the form

`tenant id, attachment generation, S3 key`.

IMO the validation & processing can run anywhere, including a third party process.

Reply:
editorial: clarify that  we are not claiming the scrubbing _can't_ run externally, just that it is not intended.